File: scripts/ghidra_extract.py
#!/usr/bin/env python3
"""Ghidra headless post-analysis script to extract decompiled functions and string xrefs.

Run via Ghidra headless:
  analyzeHeadless /tmp/ghidra_esp32_project esp32_dongle \
    -process app_code_IROM.bin \
    -noanalysis \
    -scriptPath /path/to/this/dir \
    -postScript ghidra_extract.py /path/to/output_dir
"""

# This script runs inside Ghidra's Jython environment
# pylint: disable=undefined-variable
import os

from ghidra.app.decompiler import DecompInterface, DecompileOptions
from ghidra.util.task import ConsoleTaskMonitor


def get_output_dir():
    """Get output directory from script args or default."""
    args = getScriptArgs()  # noqa: F821
    if args:
        return args[0]
    return "/tmp/ghidra_esp32_output"


def setup_decompiler(program):
    """Initialize decompiler interface."""
    decomp = DecompInterface()
    opts = DecompileOptions()
    decomp.setOptions(opts)
    decomp.openProgram(program)
    return decomp


def find_strings_containing(program, keywords):
    """Find defined strings containing any keyword, return {addr: string_value}."""
    listing = program.getListing()
    program.getMemory()
    strings = {}

    # Walk all defined data looking for strings
    data_iter = listing.getDefinedData(True)
    while data_iter.hasNext():
        d = data_iter.next()
        dt = d.getDataType()
        if "string" in dt.getName().lower():
            try:
                val = d.getValue()
                if val is not None:
                    s = str(val)
                    for kw in keywords:
                        if kw.lower() in s.lower():
                            strings[d.getAddress()] = s
                            break
            except Exception:
                pass

    return strings


def get_xrefs_to(program, addr):
    """Get all references to an address."""
    ref_mgr = program.getReferenceManager()
    return list(ref_mgr.getReferencesTo(addr))


def decompile_function_at(decomp, program, addr, monitor):
    """Decompile the function containing the given address."""
    func_mgr = program.getFunctionManager()
    func = func_mgr.getFunctionContaining(addr)
    if func is None:
        return None, None

    result = decomp.decompileFunction(func, 30, monitor)
    if result and result.depiledFunction():
        return func, result.getDecompiledFunction().getC()
    return func, None


def export_all_functions(program, decomp, monitor, output_dir):
    """Export summary of all functions."""
    func_mgr = program.getFunctionManager()
    funcs = []

    func_iter = func_mgr.getFunctions(True)
    while func_iter.hasNext():
        f = func_iter.next()
        funcs.append(
            {
                "name": f.getName(),
                "addr": str(f.getEntryPoint()),
                "size": f.getBody().getNumAddresses(),
            }
        )

    # Sort by size descending
    funcs.sort(key=lambda x: -x["size"])

    path = os.path.join(output_dir, "function_list.txt")
    with open(path, "w") as fh:
        fh.write("# All functions sorted by size (largest first)\n")
        fh.write("# Address | Size | Name\n")
        fh.write("-" * 60 + "\n")
        for f in funcs:
            fh.write("0x{:<12s} {:>8d}  {}\n".format(f["addr"], f["size"], f["name"]))
        fh.write("\nTotal functions: {}\n".format(len(funcs)))

    print("Exported {} functions to {}".format(len(funcs), path))
    return funcs


def export_string_xrefs(program, decomp, monitor, output_dir, keywords):
    """Find strings matching keywords and decompile referencing functions."""
    strings = find_strings_containing(program, keywords)
    print("Found {} matching strings".format(len(strings)))

    path = os.path.join(output_dir, "string_xrefs.txt")
    decompiled_path = os.path.join(output_dir, "decompiled_protocol.c")

    seen_funcs = set()
    decompiled_code = []

    with open(path, "w") as fh:
        fh.write("# String cross-references for protocol analysis\n")
        fh.write("# Keywords: {}\n\n".format(", ".join(keywords)))

        for addr, s in sorted(strings.items(), key=lambda x: str(x[1])):
            fh.write('\n--- String at {}: "{}"\n'.format(addr, s[:200]))
            xrefs = get_xrefs_to(program, addr)
            if not xrefs:
                fh.write("  (no references found)\n")
                continue

            for ref in xrefs:
                from_addr = ref.getFromAddress()
                func_mgr = program.getFunctionManager()
                func = func_mgr.getFunctionContaining(from_addr)
                if func:
                    func_name = func.getName()
                    func_addr = str(func.getEntryPoint())
                    fh.write("  Referenced by: {} @ {}\n".format(func_name, func_addr))

                    # Decompile if we haven't already
                    if func_addr not in seen_funcs:
                        seen_funcs.add(func_addr)
                        result = decomp.decompileFunction(func, 30, monitor)
                        if result and result.getDecompiledFunction():
                            c_code = result.getDecompiledFunction().getC()
                            if c_code:
                                decompiled_code.append(
                                    '// === {} @ {} ===\n// Referenced string: "{}"\n\n{}\n'.format(
                                        func_name, func_addr, s[:100], c_code
                                    )
                                )
                else:
                    fh.write(
                        "  Referenced from: {} (not in a function)\n".format(from_addr)
                    )

    # Write all decompiled protocol functions
    with open(decompiled_path, "w") as fh:
        fh.write(
            "// Decompiled protocol-related functions from ESP32-C3 dongle firmware\n"
        )
        fh.write(
            "// Generated by Ghidra {} headless analysis\n\n".format(
                getGhidraVersion() if "getGhidraVersion" in dir() else "12.0.3"  # noqa: F821
            )
        )
        for code in decompiled_code:
            fh.write(code)
            fh.write("\n" + "=" * 80 + "\n\n")

    print(
        "Exported {} decompiled functions to {}".format(
            len(decompiled_code), decompiled_path
        )
    )
    return seen_funcs


def export_large_functions(program, decomp, monitor, output_dir, top_n=50):
    """Decompile the N largest functions."""
    func_mgr = program.getFunctionManager()
    funcs = []

    func_iter = func_mgr.getFunctions(True)
    while func_iter.hasNext():
        f = func_iter.next()
        funcs.append(f)

    # Sort by size descending
    funcs.sort(key=lambda f: -f.getBody().getNumAddresses())

    path = os.path.join(output_dir, "decompiled_large.c")
    with open(path, "w") as fh:
        fh.write("// Top {} largest functions decompiled\n\n".format(top_n))
        for f in funcs[:top_n]:
            result = decomp.decompileFunction(f, 30, monitor)
            if result and result.getDecompiledFunction():
                c_code = result.getDecompiledFunction().getC()
                if c_code:
                    fh.write(
                        "// === {} @ {} ({} bytes) ===\n\n{}\n".format(
                            f.getName(),
                            f.getEntryPoint(),
                            f.getBody().getNumAddresses(),
                            c_code,
                        )
                    )
                    fh.write("\n" + "=" * 80 + "\n\n")

    print("Exported top {} functions to {}".format(top_n, path))


def run():
    output_dir = get_output_dir()
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    program = getCurrentProgram()  # noqa: F821
    monitor = ConsoleTaskMonitor()
    decomp = setup_decompiler(program)

    print("\n=== Ghidra ESP32-C3 Firmware Analysis ===\n")

    # 1. Export all functions
    export_all_functions(program, decomp, monitor, output_dir)

    # 2. Find protocol-related strings and decompile referencing functions
    protocol_keywords = [
        "FUNC_CODE",
        "DATA_TRANSMISSION",
        "GET_PARAM",
        "SET_PARAM",
        "HEARBEAT",
        "HEARTBEAT",
        "rs485",
        "RS485",
        "tcp_client",
        "tcp_server",
        "modbus",
        "register",
        "inverter",
        "battery",
        "soc",
        "hido-iot",
        "solarcloudsystem",
        "luxpower",
        "wifi_config",
        "ble_",
        "uart",
        "serial_port",
        "crc",
        "encrypt",
        "decrypt",
        "firmware",
        "ota_",
        "update",
    ]
    export_string_xrefs(program, decomp, monitor, output_dir, protocol_keywords)

    # 3. Decompile the 50 largest functions
    export_large_functions(program, decomp, monitor, output_dir, top_n=50)

    print("\n=== Analysis complete. Output in: {} ===".format(output_dir))


run()

File: scripts/GhidraExtract.java
// Ghidra headless post-analysis script (Java) to extract decompiled functions and string xrefs.
//
// Run via:
//   analyzeHeadless /tmp/ghidra_esp32_project esp32_dongle \
//     -process app_code_IROM.bin -noanalysis \
//     -scriptPath /path/to/scripts \
//     -postScript GhidraExtract.java /path/to/output_dir
//
//@category Analysis

import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileOptions;
import ghidra.app.decompiler.DecompileResults;
import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.*;
import ghidra.program.model.symbol.*;
import ghidra.util.task.ConsoleTaskMonitor;

import java.io.*;
import java.util.*;
import java.util.stream.*;

public class GhidraExtract extends GhidraScript {

    private static final String[] PROTOCOL_KEYWORDS = {
        "FUNC_CODE", "DATA_TRANSMISSION", "GET_PARAM", "SET_PARAM",
        "HEARBEAT", "HEARTBEAT", "rs485", "RS485",
        "tcp_client", "tcp_server", "modbus", "register",
        "inverter", "battery", "soc", "hido-iot",
        "solarcloudsystem", "luxpower", "wifi_config", "ble_",
        "uart", "serial_port", "crc", "encrypt", "decrypt",
        "firmware", "ota_", "update", "dongle",
        "charge", "discharge", "grid", "pv_", "solar",
        "mqtt", "http", "json", "parse",
        "cmd", "command", "packet", "frame", "protocol",
    };

    @Override
    public void run() throws Exception {
        String[] args = getScriptArgs();
        String outputDir = args.length > 0 ? args[0] : "/tmp/ghidra_esp32_output";
        new File(outputDir).mkdirs();

        Program program = currentProgram;
        DecompInterface decomp = new DecompInterface();
        DecompileOptions opts = new DecompileOptions();
        decomp.setOptions(opts);
        decomp.openProgram(program);

        println("\n=== Ghidra ESP32-C3 Firmware Analysis ===\n");

        // 1. Export all functions sorted by size
        exportAllFunctions(program, outputDir);

        // 2. Find protocol strings and decompile referencing functions
        exportStringXrefs(program, decomp, outputDir);

        // 3. Decompile the 50 largest functions
        exportLargeFunctions(program, decomp, outputDir, 50);

        println("\n=== Analysis complete. Output in: " + outputDir + " ===");
        decomp.dispose();
    }

    private void exportAllFunctions(Program program, String outputDir) throws Exception {
        FunctionManager funcMgr = program.getFunctionManager();
        List<Function> funcs = new ArrayList<>();
        FunctionIterator iter = funcMgr.getFunctions(true);
        while (iter.hasNext()) {
            funcs.add(iter.next());
        }
        // Sort by size descending
        funcs.sort((a, b) -> Long.compare(b.getBody().getNumAddresses(), a.getBody().getNumAddresses()));

        String path = outputDir + "/function_list.txt";
        try (PrintWriter pw = new PrintWriter(new FileWriter(path))) {
            pw.println("# All functions sorted by size (largest first)");
            pw.println("# Address      |    Size | Name");
            pw.println("-".repeat(70));
            for (Function f : funcs) {
                pw.printf("0x%-12s %8d  %s%n",
                    f.getEntryPoint().toString(),
                    f.getBody().getNumAddresses(),
                    f.getName());
            }
            pw.println("\nTotal functions: " + funcs.size());
        }
        println("Exported " + funcs.size() + " functions to " + path);
    }

    private void exportStringXrefs(Program program, DecompInterface decomp, String outputDir) throws Exception {
        Listing listing = program.getListing();
        ReferenceManager refMgr = program.getReferenceManager();
        FunctionManager funcMgr = program.getFunctionManager();

        // Find all defined strings matching keywords
        Map<Address, String> matchingStrings = new LinkedHashMap<>();
        DataIterator dataIter = listing.getDefinedData(true);
        while (dataIter.hasNext()) {
            Data d = dataIter.next();
            String dtName = d.getDataType().getName().toLowerCase();
            if (dtName.contains("string") || dtName.contains("char")) {
                Object val = d.getValue();
                if (val != null) {
                    String s = val.toString();
                    for (String kw : PROTOCOL_KEYWORDS) {
                        if (s.toLowerCase().contains(kw.toLowerCase())) {
                            matchingStrings.put(d.getAddress(), s);
                            break;
                        }
                    }
                }
            }
        }
        println("Found " + matchingStrings.size() + " matching strings");

        String xrefPath = outputDir + "/string_xrefs.txt";
        String decompPath = outputDir + "/decompiled_protocol.c";

        Set<String> seenFuncs = new HashSet<>();
        List<String> decompiled = new ArrayList<>();

        try (PrintWriter xrefPw = new PrintWriter(new FileWriter(xrefPath))) {
            xrefPw.println("# String cross-references for protocol analysis");
            xrefPw.println("# Keywords: " + String.join(", ", PROTOCOL_KEYWORDS));
            xrefPw.println();

            for (Map.Entry<Address, String> entry : matchingStrings.entrySet()) {
                Address addr = entry.getKey();
                String s = entry.getValue();
                String truncated = s.length() > 200 ? s.substring(0, 200) + "..." : s;
                xrefPw.println("\n--- String at " + addr + ": \"" + truncated + "\"");

                List<Reference> refs = new ArrayList<>();
                ReferenceIterator refIter = refMgr.getReferencesTo(addr);
                while (refIter.hasNext()) {
                    refs.add(refIter.next());
                }
                if (refs.isEmpty()) {
                    xrefPw.println("  (no references found)");
                    continue;
                }

                for (Reference ref : refs) {
                    Address fromAddr = ref.getFromAddress();
                    Function func = funcMgr.getFunctionContaining(fromAddr);
                    if (func != null) {
                        String funcName = func.getName();
                        String funcAddr = func.getEntryPoint().toString();
                        xrefPw.println("  Referenced by: " + funcName + " @ " + funcAddr);

                        if (!seenFuncs.contains(funcAddr)) {
                            seenFuncs.add(funcAddr);
                            DecompileResults result = decomp.decompileFunction(func, 30, monitor);
                            if (result != null && result.getDecompiledFunction() != null) {
                                String code = result.getDecompiledFunction().getC();
                                if (code != null) {
                                    String label = s.length() > 100 ? s.substring(0, 100) + "..." : s;
                                    decompiled.add(
                                        "// === " + funcName + " @ " + funcAddr + " ===\n" +
                                        "// Referenced string: \"" + label + "\"\n\n" +
                                        code + "\n"
                                    );
                                }
                            }
                        }
                    } else {
                        xrefPw.println("  Referenced from: " + fromAddr + " (not in a function)");
                    }
                }
            }
        }

        try (PrintWriter pw = new PrintWriter(new FileWriter(decompPath))) {
            pw.println("// Decompiled protocol-related functions from ESP32-C3 dongle firmware");
            pw.println("// Generated by Ghidra headless analysis");
            pw.println("// Total protocol functions: " + decompiled.size());
            pw.println();
            for (String code : decompiled) {
                pw.println(code);
                pw.println("=".repeat(80));
                pw.println();
            }
        }
        println("Exported " + decompiled.size() + " decompiled protocol functions to " + decompPath);
    }

    private void exportLargeFunctions(Program program, DecompInterface decomp, String outputDir, int topN) throws Exception {
        FunctionManager funcMgr = program.getFunctionManager();
        List<Function> funcs = new ArrayList<>();
        FunctionIterator iter = funcMgr.getFunctions(true);
        while (iter.hasNext()) {
            funcs.add(iter.next());
        }
        funcs.sort((a, b) -> Long.compare(b.getBody().getNumAddresses(), a.getBody().getNumAddresses()));

        String path = outputDir + "/decompiled_large.c";
        int count = 0;
        try (PrintWriter pw = new PrintWriter(new FileWriter(path))) {
            pw.println("// Top " + topN + " largest functions decompiled from ESP32-C3 dongle firmware");
            pw.println();
            for (int i = 0; i < Math.min(topN, funcs.size()); i++) {
                Function f = funcs.get(i);
                DecompileResults result = decomp.decompileFunction(f, 30, monitor);
                if (result != null && result.getDecompiledFunction() != null) {
                    String code = result.getDecompiledFunction().getC();
                    if (code != null) {
                        pw.printf("// === %s @ %s (%d bytes) ===%n%n%s%n",
                            f.getName(), f.getEntryPoint(),
                            f.getBody().getNumAddresses(), code);
                        pw.println("=".repeat(80));
                        pw.println();
                        count++;
                    }
                }
            }
        }
        println("Exported " + count + " large functions to " + path);
    }
}
